packages/grafana-data/src/dataframe/ArrayDataFrame.ts

```diff
@@ -1,118 +1,69 @@
-import { vectorToArray } from '../vector/vectorToArray';
 
-import { guessFieldTypeFromNameAndValue, toDataFrameDTO } from './processDataFrame';
-
-/** @public */
-export type ValueConverter<T = any> = (val: unknown) => T;
-
-const NOOP: ValueConverter = (v) => v;
-
-class ArrayPropertyVector<T = any> extends FunctionalVector<T> {
-  converter = NOOP;
-
-  constructor(private source: any[], private prop: string) {
-    super();
-    return makeArrayIndexableVector(this);
-  }
-
-  get length(): number {
-    return this.source.length;
-  }
-
-  get(index: number): T {
-    return this.converter(this.source[index][this.prop]);
-  }
-
-  toArray(): T[] {
-    return vectorToArray(this);
-  }
-
-  toJSON(): T[] {
-    return vectorToArray(this);
-  }
-}
+import { guessFieldTypeForField } from './processDataFrame';
 
 /**
  * The ArrayDataFrame takes an array of objects and presents it as a DataFrame
  *
- * @alpha
+ * @deprecated use arrayToDataFrame
  */
-export class ArrayDataFrame<T = any> extends FunctionalVector<T> implements DataFrame {
+export class ArrayDataFrame<T = any> implements DataFrame {
+  fields: Field[] = [];
+  length = 0;
   name?: string;
   refId?: string;
   meta?: QueryResultMeta;
 
-  fields: Field[] = [];
-  length = 0;
+  constructor(source: T[], names?: string[]) {
+    return arrayToDataFrame(source, names) as ArrayDataFrame<T>; // returns a standard DataFrame
+  }
+}
 
-  constructor(private source: T[], names?: string[]) {
-    super();
+/**
+ * arrayToDataFrame will convert any array into a DataFrame
+ *
+ * @public
+ */
+export function arrayToDataFrame(source: any[], names?: string[]): DataFrame {
+  const df: DataFrame = {
+    fields: [],
+    length: source.length,
+  };
+  if (!source?.length) {
+    return df;
+  }
 
-    this.length = source.length;
-    const first: any = source.length ? source[0] : {};
-    if (names) {
-      this.fields = names.map((name) => {
-        return {
+  if (names) {
+    for (const name of names) {
+      df.fields.push(
+        makeFieldFromValues(
           name,
-          type: guessFieldTypeFromNameAndValue(name, first[name]),
-          config: {},
-          values: new ArrayPropertyVector(source, name),
-        };
-      });
-    } else {
-      this.setFieldsFromObject(first);
+          source.map((v) => v[name])
+        )
+      );
     }
-    return makeArrayIndexableVector(this);
+    return df;
   }
 
-  /**
-   * Add a field for each property in the object.  This will guess the type
-   */
-  setFieldsFromObject(obj: Record<string, unknown>) {
-    this.fields = Object.keys(obj).map((name) => {
-      return {
-        name,
-        type: guessFieldTypeFromNameAndValue(name, obj[name]),
-        config: {},
-        values: new ArrayPropertyVector(this.source, name),
-      };
-    });
-  }
-
-  /**
-   * Configure how the object property is passed to the data frame
-   */
-  setFieldType(name: string, type: FieldType, converter?: ValueConverter): Field {
```

Comment:
removing this is a breaking change -- but the behavior is so weird I think that is OK

Reply:
If anyone else out there trips over this issue - just use the `arrayToDataFrame` method to initialize the `DataFrame` and set each field's type after the `DataFrame` has been initialized (it does its best to determine the appropriate field type but doesn't get everything).